### ISSUE TITLE
cppcoreguidelines-rvalue-reference-param-not-moved false positive in the presence of an unresolved include

### ISSUE THREAD

**Issue by igor-akhmetov** [llvm/llvm-project#197569]
In this snippet:
```
#include <utility>
#include <xxx>

struct X {};

void foo(X&& input)
{
    X copy(std::move(input));
}
```
clang-tidy reports cppcoreguidelines-rvalue-reference-param-not-moved on the `input` parameter, see https://gcc.godbolt.org/z/5rnjTT5aY.
If I remove the unresolved include, the check goes away.

**Comment by vbvictor** [comment]
Please see: https://clang.llvm.org/extra/clang-tidy/index.html

> Clang compiler errors (such as syntax errors, semantic errors, or other failures that prevent Clang from compiling the code) are reported with the check name clang-diagnostic-error. These represent fundamental compilation failures that must be fixed before clang-tidy can perform its analysis. Unlike other diagnostics, clang-diagnostic-error cannot be disabled, as clang-tidy requires valid code to function.

"clang-tidy requires valid code to function"

We will hopefully soon disable analysis for broken TUs https://github.com/llvm/llvm-project/issues/196733 (which is your case).

**Comment by igor-akhmetov** [comment]
I'm aware of this principle, but don't you find it strange that an unresolved include affects a completely unrelated clang-tidy check?

An opt-out flag is required for https://github.com/llvm/llvm-project/issues/196733. Clang-tidy is used in some IDEs, and having errors in a TU is a usual scenario there. Yes, I know that clangd is preferred for this use case.

**Comment by vbvictor** [comment]
> I'm aware of this principle, but don't you find it strange that an unresolved include affects a completely unrelated clang-tidy check?

If you look at AST https://gcc.godbolt.org/z/j3EbdjKzr, you will see that it contains RecoveryExpr so the check can't find `callExpr` (which will be "move" we want to find):

```
`-FunctionDecl <line:6:1, line:9:1> line:6:6 foo 'void (X &&)'
  |-ParmVarDecl <col:10, col:14> col:14 used input 'X &&'
  `-CompoundStmt <line:7:1, line:9:1>
    `-DeclStmt <line:8:5, col:29>
      `-VarDecl <col:5, col:28> col:7 copy 'X' callinit
        `-ParenListExpr <col:11, col:28> 'NULL TYPE' contains-errors
          `-RecoveryExpr <col:12, col:27> 'typename std::remove_reference<_Tp>::type' contains-errors lvalue
            |-UnresolvedLookupExpr <col:12, col:17> '<overloaded function type>' lvalue (no ADL) = 'move' 0x3bc50588
            `-DeclRefExpr <col:22> 'X' lvalue ParmVar 0x3bce0f78 'input' 'X &&'
```

If you remove broken import different AST is build that has `CallExpr`:

```
`-FunctionDecl <line:5:1, line:8:1> line:5:6 foo 'void (X &&)'
  |-ParmVarDecl <col:10, col:14> col:14 used input 'X &&'
  `-CompoundStmt <line:6:1, line:8:1>
    `-DeclStmt <line:7:5, col:29>
      `-VarDecl <col:5, col:28> col:7 copy 'X' callinit
        `-CXXConstructExpr <col:7, col:28> 'X' 'void (X &&) noexcept'
          `-CallExpr <col:12, col:27> 'typename std::remove_reference<X &>::type':'X' xvalue
            |-ImplicitCastExpr <col:12, col:17> 'typename std::remove_reference<X &>::type &&(*)(X &) noexcept' <BuiltinFnToFnPtr>
            | `-DeclRefExpr <col:12, col:17> '<builtin fn type>' Function 0x2f1cb5b0 'move' 'typename std::remove_reference<X &>::type &&(X &) noexcept' (FunctionTemplate 0x2f13a368 'move')
            |   `-NestedNameSpecifier Namespace 0x2f1c2880 'std'
            `-DeclRefExpr <col:22> 'X' lvalue ParmVar 0x2f1cab48 'input' 'X &&'
```

So that is why we don't want to run analysis on any broken TU.
From my experience, broken import messes up the whole file (I've seen other checks producing FP as well).

**Comment by igor-akhmetov** [comment]
Right, so the question is why does an unresolved include directive cause lookup to fail? `<utility>` is still included, so the call to `move` should be resolved.

**Comment by vbvictor** [comment]
> Right, so the question is why does an unresolved include directive cause lookup to fail? `<utility>` is still included, so the call to `move` should be resolved.

Yes, the question can go into `clang` area about failed lookup and `RecoveryExpr` in (possibly) unrelated code.
I suspect this issue may not be the only one where broken include messes some other part of AST.
